Port over to use scraped

chrismytton opened the pull request:
This updates this scraper to use `Scraped` subclasses.

# Notes to reviewer

I would have ideally liked to end up with fewer `Member` classes, but the various pages are _just_ different enough to make this awkward. I think what's here is definitely an improvement though.

This also shows https://github.com/everypolitician/scraped/pull/32 in action. It feels much more natural to decorate the individual classes, rather than the response, so I think that's good to go though now.

# Notes to merger

- [x] Update to use `scraped@master` once https://github.com/everypolitician/scraped/pull/32 has been merged

chrismytton committed:
Update to latest Ruby version

The `scraped` gem uses keyword arguments, so requires Ruby 2.1+.

chrismytton committed:
Add scraped dependency

chrismytton committed:
Add require_all gem

chrismytton committed:
Run rubocop -a on Gemfile

chrismytton committed:
Pull Parser logic into separate classes for each page type

As a first step this just moves the logic for the different types of
page into their own classes.

chrismytton committed:
Remove Parser class

Just use the Page classes directly now.

chrismytton committed:
Remove unused constant WARNED

This isn't being used anywhere (and never was as far as I can tell).

chrismytton committed:
Use the UnspanAllTables decorator on the response

This goes through all the tables on the page and removes the colspan and
rowspan attributes so that the table can just be treated normally,
rather than having to jump between rows and columns to get the value
you're after.

chrismytton committed:
Add RemoveUnwantedTable class

This logic was previously happening inline before pulling out the wanted
tables. Instead we now want to do this in a decorator on the specific
class.

chrismytton committed:
Switch to scraped class-level-decorators branch

This will allow us to decorate the HTML at the individual class level,
which will hopefully mean we can simplify the table unspanner.

chrismytton committed:
Fix requires in unspanned_table.rb

chrismytton committed:
Extract Member classe for various page types

These are currently all a bit different from each other, but the plan is
to use decorators to smooth over the inconsistencies.

chrismytton committed:
Remove global decorators

We want to decorate individual classes, rather than do it at the top
level.

chrismytton committed:
Update to latest version of scraped

This version has a fix to stop mutating the decorator config so that the
class' decorators can be called multiple times.

chrismytton committed:
Extract PartyInformation class

This class handles mapping a party name to a name + id combo.

chrismytton committed:
Don't mutate original table in UnspannedTable

Rather than transform/mutate the original nokogiri element we instead
just return a string representing the children of the transformed table,
which is then set on the original table in UnspanAllTables.

chrismytton committed:
Simplify TermFourColumnPage using decorators

This uses the unspan all tables decorator to simplify this class and fix
the area and party fields so that they return the correct values for all
people.

chrismytton committed:
Extract member class in TermSinglePartyPage

This uses the UnspanAllTables decorator to simplify the code for
extracting a member on this page type.

chrismytton committed:
Remove pry breakpoint

chrismytton committed:
Update TermThreeColumnPage to use UnspanAllTables decorator

tmtmtmtm commented on lib/party_information.rb:
```diff
@@ -0,0 +1,66 @@
+        else
+          party_name
+        end
+      end
```
This is a little bit complex, and it looks like the complexity is almost entirely down to the memoization, which I suspect is just premature optimisation. Without it this could just be:

 ```ruby
  return party_name unless party_name.include? '→'
  party_name.split(/\s*→\s*/).first
```

chrismytton replied:
Good point, simplified in https://github.com/everypolitician-scrapers/turkey-tbmm-wikipedia/pull/3/commits/b906e511821d66dd7ebda69afca51cdb3f29adbd

tmtmtmtm commented on lib/remove_unwanted_table.rb:
```diff
@@ -0,0 +1,7 @@
+class RemoveUnwantedTable < Scraped::Response::Decorator
```
I think the name of this class is a bit too general, when this is a very specific thing to be doing.

chrismytton replied:
Yep, good catch. It turns out we now longer need this now we're auto-detecting tables, since we now just ignore tables that we don't handle, so I've removed it altogether.

tmtmtmtm reviewed (changes requested):
I don't have time for a full review right now I'm afraid, but a couple of drive-by comments…

chrismytton committed:
Use UnspanAllTables everywhere

This is more consistent, rather than having some places that have
unspanned tables and some that don't.

chrismytton committed:
Remove unused variables

chrismytton committed:
Move term table page classes into term_table namespace

These are really representing different kinds of table, not different
kinds of page, so putting them into this namespace makes that clear. The
next thing to change will be to actually make these table level.

chrismytton committed:
Add TermTable::Base class

This allows us to define the :members field once and then have
subclasses tell use which headers to look for an then define a `Member`
for that table type.

I think this is a step in the right direction but the individual table
classes are now essentially just representing rows, so we can move those
out of the way as well.

chrismytton committed:
Rename TermTable -> TermTableRow

These are going to be representing the individual rows, as there wasn't
any logic relating to the tables in these classes.

chrismytton committed:
Remove Member classes

Just use the various different types as row classes directly.

chrismytton committed:
Add classes representing term pages and tables

This allows us to auto-detect the correct row class for each table that
we're dealing with.

chrismytton committed:
Switch to field_serializer master branch

This has the fix for broken inheritence that hasn't been released yet.

chrismytton committed:
Use a method for party_name in rows

We return a :party and :party_id field from the superclass, this is the
cleaned up version of `party_name`, so we just have that as a method
that can be used by the base class.

chrismytton committed:
Add methods to TermTableRow::Base for other fields

These are the fields that are common between all rows.

chrismytton committed:
Simplify members field

chrismytton committed:
Return full members rows

We shouldn't assume that people want to deal with a hash, so pass back
the full instance.

chrismytton committed:
Switch to auto-detection of row classes

chrismytton committed:
Remove unused RemoveUnwantedTable class

Now that we're auto-detecting the correct class for given tables we no
longer need this as we just ignore tables that we can't process.

chrismytton committed:
Simplify PartyInformation#party method

The memoization was just complicating things and wasn't really
necessary.

chrismytton commented:
This is now auto-detecting the correct row-class to use based on the column headings for the table _after unspanning_. This has dramatically simplified the code in `scraper.rb`, but has meant introducing a base class for the rows. I think the base class is OK for now, but I'm open to suggestions to better alternatives.

chrismytton commented:
Going to try doing a rebuild of everypolitician-data with the output from this to see if it's correctly fixed the people that were going missing.

chrismytton committed:
Use field_serializer from RubyGems

chrismytton committed:
Simplify requires in scraper.rb

chrismytton committed:
Move row validity check into row class

Tell, don't ask.

chrismytton committed:
Allow expected_headers to include regexes

We need to be able to do a partial match on a header for term 24.
The `#match` method is implemented by both `String` and `Regexp`, so we
can use just call match and it should do the right thing.

chrismytton committed:
Use a regex to match odd term 24 table

There is a table on the term 24 page that has four columns, but only two
are useful, and they are the same two that are useful in the
ByAreaTwocol class. So we use a regex to match the column headings so it
works for both tables.

chrismytton committed:
Add a temporary fix for term 18

I've submitted a change to fix a problem with a rowspan that is one
shorter than it needs to be, but that's pending review currently. So
just add a temporary workaround for now.

chrismytton committed:
Revert "Add a temporary fix for term 18"

This reverts commit 45d4075a138d169e8812089d5de9d010c275aa61.

You can actually link directly to a revision of a page, so we'll do that
instead of using this hacky approach.

chrismytton commented:
> Wouldn't you be better just scraping the non-approved version of the page instead for now, and then switching back again once it's approved?
> 
> We've done this several times before — e.g. 9afd4ca

❤️  I didn't know you could do that, much simpler!

chrismytton committed:
Use unstable urls for terms that have been edited

I've had to fix the markup in a few places, so link to the unstable
versions of those pages until the changes are approved.

chrismytton commented:
Woo, that's all working nicely with the overrides in https://github.com/everypolitician-scrapers/turkey-tbmm-wikipedia/pull/3/commits/789c6812e520c68deb563444be61efb23f9b01a7 for the pages I've had to edit.

I've opened https://github.com/everypolitician/everypolitician-data/pull/21814 as a test that the data is doing the right thing etc.

So I think this is ready for review! 👀

chrismytton committed:
Update scraped to a valid version

This was previously pointing to an orphaned commit.

ondenman reviewed (commented):
Is there any need to add 'require_all'? I see that Scraped is already requiring it.

Ah... I see you've removed it anyway. Never mind!

ondenman reviewed (approved):
All looks good to me!

chrismytton committed:
Update to scraped@master

chrismytton committed:
Add table_unspanner dependency

chrismytton committed:
Switch to using table_unspanner gem

chrismytton committed:
Remove local version of table unspanner

chrismytton committed:
Update to latest scraped

chrismytton committed:
Add decorator to unspan all tables on the page

This is currently producing nokogiri warnings, I don't think they're
anything to worry about, but I haven't dug into them at all.

chrismytton committed:
Use the UnspanAllTables decorator

tmtmtmtm reviewed (approved):
There are a few things that I'd like to tweak on this, but as we don't have a functional scraper at all at the minute, I'm going to merge this and we can evolve from there.
